### DISCUSSION TITLE
Documentation for using Rhino on Android

### DISCUSSION THREAD

**Discussion by p-bakker** [mozilla/rhino#1162] [Q&A]
Hi all,

I'm looking for input for documentation on using Rhino on Android. 

It looks like Rhino is being used on Android by quite a few people, but the involved in maintaining Rhino aren't among that group.

I'd like to add some documentation about using Rhno on Android, things like use-cases, tips & tricks, requirements, pitfalls, known issues, things to look out for, limitations etc. but not using Rhino on Android myself, I have no insight in any of these items.

So, hereby a request to all devs that use Rhino on Android to contribute. 

If you're willing to write the documentation that would be great, lemme know in a comment. If you just have some bits of info to share that you thing ought to be documented, also great, just add them as comments as well!

Note that we're tracking Android related issues using the [Android label](https://github.com/mozilla/rhino/issues?q=is%3Aissue+label%3AAndroid+)

**Comment by p-bakker** [comment]
Tagging GitHub users here that have been raising or commenting on some of our Android-related issues in the recent or somewhat distant past in the hopes of interesting you in sharing your thoughts on this discussion or maybe even contributing: @skbeh @ylk2534246654 @b3nn0 @msridhar @archethic @currydou @PragyaSingla @weiyang94 @JayNewstrom @TheBrokenRail @F43nd1r @AlexTrotsenko

**Comment by p-bakker** [comment]
Already identified things to (possibly) document related to Rhino on Android:

1. https://github.com/F43nd1r/rhino-android
2. Potential Secure XML Parsing issues, see #1158 & Context.FEATURE_ENABLE_XML_SECURE_PARSING
3. How to properly use Proguard for Rhino on Android
4. Rhino's ByteCode generation doesn't generate Andoid-compatible bytecode, so Rhino can only be used in Interpreted Mode (`Context.setOptimizationLevel(-1)`) and the JavaAdapter is not supported

**Reply by F43nd1r** [reply]
My wrapper (1.) is a fix/workaround for bytecode issues and with that JavaAdapter usage (4.)

**Reply by p-bakker** [reply]
@F43nd1r thnx for your input.

Given your comment https://github.com/mozilla/rhino/issues/1152#issuecomment-1017749680: are you still supporting your wrapper! And is it tied to a specific Rhino version or does it work with any Rhino version?

**Reply by F43nd1r** [reply]
> are you still supporting your wrapper?

Yes, I'm just not a user myself anymore. So I'll only react to issues and feature requests, not drive development myself.

> And is it tied to a specific Rhino version or does it work with any Rhino version?

Compatible with most™ rhino versions. I started from Rhino 1.7.7, current build is with 1.7.13. As long as rhino stays binary compatible it should work with future and past versions.

**Reply by p-bakker** [reply]
Kk, great!

**Reply by p-bakker** [reply]
@F43nd1r  one more question: what do you think of somehow making your wrapper part of the Rhino project, so we could offer it as a first class citizen right here in this repo as an Android-specific artifact?

Not sure what would be involved and what the maintainers of the rhino project think about this, so just thinking out loud, but to me it sorta makes sense

**Reply by F43nd1r** [reply]
Sure, but if we're talking first class android support in Rhino I'd hope for a direct to android compiler instead.
Currently:
```
javascript --rhino--> java bytecode
javascript --rhino--> java bytecode --dalvik-dx--> android bytecode
```
What I'd like to see:
```
javascript --rhino-jvm--> java bytecode
javascript --rhino-android--> android bytecode
```
I already mentioned that in my issues from 2016, but at that point nobody seemed to be interested in the necessary refactorings surrounding classfilewriter.
